### ISSUE TITLE
tf.linalg.LinearOperatorLowRankUpdate gradient computation causes segmentation fault with incompatible gradient shape

### ISSUE THREAD

**Issue by TTZTTZ1** [tensorflow/tensorflow#125502]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

tf2.21.0

### Custom code

Yes

### OS platform and distribution

Ubuntu 22.04.5 LTS

### Mobile device

No

### Python version

Python 3.10.20

### Bazel version

N/A

### GCC/compiler version

N/A

### CUDA/cuDNN version

CUDA 13.0 cuDNN 9.x

### GPU model and memory

NVIDIA GeForce RTX 5090 32GB

### Current behavior?

`tf.linalg.LinearOperatorLowRankUpdate` causes a segmentation fault during gradient computation when an incompatible upstream gradient shape is provided.

Instead of raising a Python-level exception, the process terminates with a segmentation fault.
TensorFlow should validate the gradient shape and raise a clear Python-level exception (such as ValueError or InvalidArgumentError) instead of causing a segmentation fault.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

base = tf.linalg.LinearOperatorIdentity(num_rows=3)

u = tf.random.normal([3, 3])

g = tf.constant([4.0])

with tf.GradientTape() as tape:
    result = tf.linalg.LinearOperatorLowRankUpdate(
        base,
        u,
        u
    ).matmul(base)

tape.watch(result)

tape.gradient(result, [u], [g])
```

### Relevant log output

```shell
Segmentation fault
```

**Comment by vishwakt** [comment]
@Kayyuri 
Reproduced on 2.21.0 (CPU). Correcting my earlier comment: the root cause is not the target type, and not the gradient's *shape*. It is a count mismatch.

**The trigger is the number of gradients, not their shape.** With a correctly shaped `output_gradients` this segfaults on every run:

```python
import tensorflow as tf
base = tf.linalg.LinearOperatorIdentity(num_rows=3)
u = tf.random.normal([3, 3])
with tf.GradientTape() as tape:
    result = tf.linalg.LinearOperatorLowRankUpdate(base, u, u).matmul(base)
tape.gradient(result, [u], [tf.ones([3, 3])])   # SIGSEGV, 5/5 runs
```

The originally reported script segfaults only about 40% of the time and otherwise raises `InvalidArgumentError`, so running it once can easily look like it does not reproduce.

**Why.** `matmul(base)` is called with a `LinearOperator`, so it returns a composed `LinearOperatorLowRankUpdate`, which is a `CompositeTensor`. In `GradientTape.gradient`, `composite_tensor_gradient.get_flat_tensors_for_gradients` expands that target into **3** component tensors, while `output_gradients` is flattened separately and stays at **1**. Nothing checks that the counts agree, so the tape indexes `output_gradients` per target and reads out of bounds for targets 2 and 3.

That explains every observation, 5 runs each:

| variant | result |
| --- | --- |
| 1 gradient, 3 flat targets | SEGV 3/5 |
| 2 gradients, 3 flat targets | SEGV 3/5 |
| **3 gradients, 3 flat targets** | **ok 5/5** |
| no `output_gradients` | ok 5/5 |
| `matmul(tf.eye(3))` or `to_dense()` (returns a real Tensor) | ok 5/5 |

Supplying three correctly shaped gradients makes it work perfectly, which confirms the count is the issue. The non-determinism follows from the out-of-bounds read returning whatever happens to be in memory, and `faulthandler` puts the fault at the `isinstance` generator in `_aggregate_grads` (`backprop.py:599`), which is where the corruption surfaces rather than where it originates.

For scope: `RaggedTensor` targets expand to a single gradient tensor, so no mismatch is possible, and `SparseTensor` targets raise a clean `ValueError` earlier. `LinearOperator` appears to be the reachable case today.

Expected behavior is a `ValueError` reporting the mismatch instead of a segfault. I have a fix that adds that check and will send a PR.

**Comment by Kayyuri** [comment]
Hi @TTZTTZ1, Thank you for reporting this issue.
A fix for this issue has been proposed in this [PR](https://github.com/tensorflow/tensorflow/pull/125589). Once it is merged, this issue will be resolved.

Many thanks to @vishwakt for working on the fix contributions.
